source/slang/slang-ir-specialize.cpp

```diff
@@ -2440,44 +2474,57 @@ struct SpecializationContext
         }
         if (elementCount == 0)
         {
-            auto resultValuePack = builder.emitMakeValuePack(0, (IRInst*const*)nullptr);
-            expandInst->replaceUsesWith(resultValuePack);
+            auto resultPack = makeSpecializedPack(builder, expandInst->getDataType(), elements.getArrayView());
+            expandInst->replaceUsesWith(resultPack);
             expandInst->removeAndDeallocate();
-            addUsersToWorkList(resultValuePack);
+            addUsersToWorkList(resultPack);
             return true;
         }
+        
+        bool isMultiBlock = as<IRYield>(expandInst->getFirstBlock()->getTerminator()) == nullptr;
 
         for (UInt i = 0; i < elementCount; i++)
         {
             IRCloneEnv cloneEnv;
-            IRBlock* firstBlock = nullptr;
             IRBuilder subBuilder = builder;
-            for (auto childBlock : expandInst->getBlocks())
+            IRBlock* mergeBlock = nullptr;
+            if (isMultiBlock)
             {
-                auto newBlock = subBuilder.emitBlock();
-                if (!firstBlock)
-                    firstBlock = newBlock;
-                cloneEnv.mapOldValToNew[childBlock] = newBlock;
+                IRBlock* firstBlock = nullptr;    
+                for (auto childBlock : expandInst->getBlocks())
+                {
+                    auto newBlock = subBuilder.emitBlock();
```

Comment:
Yong's note: we want to make sure that if the body of `expand` just contains one block, we can just unroll the expand directly into the enclosing block instead of creating one new block for each unrollled iteration. This simplifies the resulting code, and is necessary for types because types may be used in a place where blocks aren't allowed to exist.